### ISSUE TITLE
Minimal build option for CMake

### ISSUE THREAD

**Issue by p-pautov** [grpc/grpc#36023]
### Is your feature request related to a problem? Please describe.
CMake build misses some useful toggles for creating minimal binary, which are available in Bazel, e.g. `grpc_no_xds`, `grpc_no_rls`, etc. These options also allow to remove some dependencies like RE2.

### Describe the solution you'd like
Ideally, there should be CMake option, say `gRPC_MINIMAL_BUILD`, which would disable all those extra features and also disable corresponding CMake level dependencies.

**Comment by veblush** [comment]
Yeah, it's a good idea to have a way to have a minimal gRPC. But I'm not sure having an option like bazel is a good approach, though. Maybe we need another build target like `grpc_core` or something.

**Comment by p-pautov** [comment]
I think, dedicated target could work as long as building minimal gRPC won't require installation of "non-minimal" dependencies (like RE2). Though, if gRPC is used indirectly, say via some other LibA, switching to minimal build would require changes in LibA's CMakeLists.txt.

**Comment by igxactly** [comment]
is there a room for contribution? if there is, i'd like to put a PR for that.

**Comment by MrMoose** [comment]
> is there a room for contribution? if there is, i'd like to put a PR for that.

Splendid! I would assume the tag "help wanted" amounts to an invitation for a PR.

**Comment by psi-sue** [comment]
For anyone who want a quick fix, the following worked for me on 1.60

```
--- a/BUILD
+++ b/BUILD
@@ -605,7 +605,7 @@ grpc_cc_library(
     ],
     defines = select({
         ":grpc_no_xds": ["GRPC_NO_XDS"],
-        "//conditions:default": [],
+        "//conditions:default": ["GRPC_NO_XDS"],
     }),
     external_deps = [
         "absl/base:core_headers",
@@ -615,7 +615,7 @@ grpc_cc_library(
     select_deps = [
         {
             ":grpc_no_xds": [],
-            "//conditions:default": GRPC_XDS_TARGETS,
+            "//conditions:default": [],
         },
     ],
     tags = [
@@ -886,8 +886,6 @@ grpc_cc_library(
         {
             ":grpc_no_xds": [],
             "//conditions:default": [
-                "grpc++_xds_client",
-                "grpc++_xds_server",
             ],
         },
         {
@@ -1482,7 +1480,7 @@ grpc_cc_library(
         "//src/core:lib/iomgr/event_engine_shims/tcp_client.h",
     ],
     defines = select({
-        "systemd": ["HAVE_LIBSYSTEMD"],
+        "systemd": [],
         "//conditions:default": [],
     }),
     external_deps = [
@@ -1507,7 +1505,7 @@ grpc_cc_library(
     ],
     language = "c++",
     linkopts = select({
-        "systemd": ["-lsystemd"],
+        "systemd": [],
         "//conditions:default": [],
     }),
     public_hdrs = GRPC_PUBLIC_HDRS + GRPC_PUBLIC_EVENT_ENGINE_HDRS,
@@ -2263,7 +2261,7 @@ grpc_cc_library(
     hdrs = [],
     defines = select({
         ":grpc_no_xds": ["GRPC_NO_XDS"],
-        "//conditions:default": [],
+        "//conditions:default": ["GRPC_NO_XDS"],
     }),
     external_deps = [
         "absl/memory",
@@ -2274,7 +2272,7 @@ grpc_cc_library(
     ],
     select_deps = [{
         ":grpc_no_xds": [],
-        "//conditions:default": ["//:grpcpp_csds"],
+        "//conditions:default": [],
     }],
     deps = [
         "gpr",

--- a/templates/CMakeLists.txt.template
+++ b/templates/CMakeLists.txt.template
@@ -811,6 +811,9 @@
   % endif
   )
 
+  target_compile_definitions(${lib.name} PRIVATE GRPC_NO_RLS)
+  target_compile_definitions(${lib.name} PRIVATE GRPC_NO_XDS)
+
   if(WIN32 AND MSVC)
     set_target_properties(${lib.name} PROPERTIES COMPILE_PDB_NAME "${lib.name}"
       COMPILE_PDB_OUTPUT_DIRECTORY <%text>"${CMAKE_BINARY_DIR}</%text>"
```
Then run `./tools/buildgen/generate_projects.sh`

This will disable rls and xds feature for CMake (breaks other builds because GRPC_NO_RLS and GRPC_NO_XDS will not be defined). I think adding a flag to CMake with current architecture is very hard to do, because Bazel is the single source of truth that knows about dependencies and it's really difficult to convince it to share them. (Bazel creates an XML with `bazel query` which is just an approximation (!) (I guess because cquery does not support XML output), this XML is parsed in python, then turned into a yaml, which is used to generate CMake from a template..). Bazel was not designed to be a build generator like CMake was, but it is used exactly in that way here which creates this situation.

It's a bit strange that XDS is disabled on the code and also on the build level, but I'm not really familiar with grpc to judge.

**Comment by om7057** [comment]
Hi @veblush, I'd like to put up a PR for this if you are open to it!

As @p-pautov pointed out, creating a completely separate grpc_core target would unfortunately break downstream CMake libraries that explicitly look to link against gRPC::grpc++. The cleanest way to handle this while preserving target compatibility is to introduce a global CMake toggle.

My proposed implementation:

Introduce option(gRPC_BUILD_MINIMAL "Build minimal gRPC without XDS/RLS" OFF) in templates/CMakeLists.txt.template.
Within the template, conditionally filter out the XDS/RLS dependencies (and their heavy transitives like re2) when the minimal build is enabled.
Inject the GRPC_NO_XDS and GRPC_NO_RLS compiler definitions into the targets so the C++ preprocessor drops the code (matching Bazel's behavior).
Run the buildgen scripts to produce the final CMakeLists.txt.
This provides the flexibility CMake users need without diverging from the Bazel source of truth. Could you assign this to me so I can get started on the PR?
